TFNetworkLayer.py

```diff
@@ -1043,53 +1047,237 @@ def batch_norm(self, data,
+        x = data.get_placeholder_merge_axes(axes=axes, keep_dims=True)
+
+      opts = dict(zip(axes, [-1] * len(axes)))
+      variables_shape = data.get_bc_shape(opts=opts)
+      variables_shape_flat = reduce(mul, variables_shape, 1)
+
+      if sample_mean is None:
+        with self.var_creation_scope():
+          moving_mean = self.add_param(tf.Variable(
+            initial_value=tf.zeros(variables_shape_flat),
+            name="%s_%s_mean" % (self.name, data.name),
+            trainable=True))
+      else:
+        moving_mean = sample_mean
+      if sample_variance is None:
+        with self.var_creation_scope():
+          moving_variance = self.add_param(tf.Variable(
+            initial_value=tf.ones(variables_shape_flat),
+            name="%s_%s_variance" % (self.name, data.name),
+            trainable=True))
+      else:
+        moving_variance = sample_variance
+      if use_std:
+        if gamma is None:
+          with self.var_creation_scope():
+            gamma = self.add_param(tf.Variable(
+              initial_value=tf.ones(variables_shape_flat),
+              name="%s_%s_gamma" % (self.name, data.name),
+              trainable=True))
+      if use_shift:
+        if beta is None:
+          with self.var_creation_scope():
+            beta = self.add_param(tf.Variable(
+              initial_value=tf.zeros(variables_shape_flat),
+              name="%s_%s_beta" % (self.name, data.name),
+              trainable=True))
+      data_format = "NCHW" if data.is_batch_feature_major else "NHWC"
+
+      def _fused_batch_norm_training():
+        return tf.nn.fused_batch_norm(
+          x,
+          gamma,
+          beta,
+          epsilon=epsilon,
+          data_format=data_format)
+
+      def _fused_batch_norm_inference():
+        return tf.nn.fused_batch_norm(
+          x,
+          gamma,
+          beta,
+          mean=moving_mean,
+          variance=moving_variance,
+          epsilon=epsilon,
+          is_training=False,
+          data_format=data_format)
+      output, mean, variance = tf.cond(
+        tf.cast(self.network.train_flag, tf.bool),
+        true_fn=_fused_batch_norm_training,
+        false_fn=_fused_batch_norm_inference)
+      # Remove Bessel's correction to be consistent with non-fused batch norm.
+      sample_size = tf.cast(
+        tf.size(x) / tf.size(variance), variance.dtype)
+      factor = (sample_size - tf.cast(1.0, variance.dtype)) / sample_size
+      variance *= factor
+
+      # Controlling updates in training and inference modes via momentum
+      momentum_mean = tf.cast(self.network.train_flag, tf.float32) * momentum
+      momentum_variance = tf.cast(self.network.train_flag, tf.float32) * momentum
+      # Add custom updates for moving_mean and moving_variance
+      mean_update = CustomUpdateExpAverage(average=mean, alpha=momentum_mean)
+      variance_update = CustomUpdateExpAverage(average=variance, alpha=momentum_variance)
```

Comment:
Why do you have the updating here different than in the other function? (Here `CustomUpdateExpAverage`, in the other it is `tf.assign_add` directly.) It should be the same.

Reply:
Also, please don't duplicate the code. Use the same code in both cases. Simplify this. I think you should not have this in an extra function.

Reply:
`tf.assign_add` works in the old batch_norm because the variables are than added to the computatin graph via `use_sample`:
https://github.com/rwth-i6/returnn/blob/ddc40b9ac02433b6fc16a8f71344a27b726024dd/TFNetworkLayer.py#L1136

In case of fused version, all computations are done inside `tf.nn.fused_batch_norm`. And later updates can't be added to the graph. So, it's done via `CustomUpdateExpAverage`.